sql: adding "NOT NULL" column to table without specifying a default leads to a panic

Given a brand new single node:

```shell
➜  cockroach-beta-20160330.darwin-10.9-amd64  ./cockroach start --insecure
build:     beta-20160330 @ 2016/03/30 14:58:36 (go1.6)
admin:     http://scott-2.local:8080
sql:       postgresql://root@scott-2.local:26257?sslmode=disable
logs:      cockroach-data/logs
store[0]:  path=cockroach-data
```

The following script causes a node panic:

```sql
CREATE DATABASE test;
SET DATABASE = test;
CREATE TABLE test (v DECIMAL);
INSERT INTO test VALUES (DECIMAL '1.0');
ALTER TABLE test ADD COLUMN v2 DECIMAL NOT NULL;
SELECT * FROM test;
```

Stack trace:
```
panic: Non-nullable column with no value!

goroutine 136 [running]:
panic(0x4e7c220, 0xc820d79680)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql.(*scanNode).maybeOutputRow(0xc820943400, 0xc820d79601)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/scan.go:547 +0x291
github.com/cockroachdb/cockroach/sql.(*scanNode).Next(0xc820943400, 0x7e0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/scan.go:157 +0x92
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc820938fc0, 0xc8201c1ef0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/select.go:108 +0x53
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc820514000, 0x7492b78, 0xc820d1cfc0, 0xc820747078, 0xece8e7232, 0xc81ba2ef85, 0x5d4b540, 0x1, 0x0, 0x0, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:894 +0x353
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc820514000, 0x7492b78, 0xc820d1cfc0, 0xc820747078, 0xc820d70101, 0x1440c8e9f6ec41b7, 0x0, 0xc820747028, 0x0, 0x0, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:778 +0x58d
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc820514000, 0xc820d785b0, 0x1, 0x1, 0xc820747078, 0xc820747028, 0x1e0101, 0x0, 0x0, 0x0, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:556 +0x630
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc820514000, 0xc820747078, 0x1, 0xc820747028, 0xc82067fb80, 0xc820d785b0, 0x1, 0x1, 0xc820911390, 0xc8209113a8, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:488 +0x147
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc8208106e0, 0xc82067fb80, 0x4)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:401 +0x325
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc8208106e0, 0x5310101, 0x1440c8e9f6ec0e27, 0x0, 0xc8209114e8, 0x0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/client/txn.go:531 +0x410
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc820514000, 0x7416210, 0xc82094a180, 0xc820747000, 0xc820a0c0a4, 0x13, 0x0, 0x0, 0x0, 0x745cd00)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:405 +0x892
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820514000, 0x7416210, 0xc82094a180, 0xc820747000, 0xc820a0c0a4, 0x13, 0x0, 0x0, 0x0, 0x0, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/executor.go:295 +0xe2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8202af200, 0x7416210, 0xc82094a180, 0xc820a0c0a4, 0x13, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:598 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc8202af200, 0x7416210, 0xc82094a180, 0xc8202af228, 0x0, 0x0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:308 +0xf2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8202af200, 0x0, 0x0, 0x0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:268 +0x109b
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820368358, 0x9200028, 0xc82075e210, 0x0, 0x0)
	/ext-go/1/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:159 +0x929
github.com/cockroachdb/cockroach/server.(*Server).Start.func4.1(0x9200028, 0xc82075e210)
	/ext-go/1/src/github.com/cockroachdb/cockroach/server/server.go:300 +0x4b
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc820308100, 0x9200028, 0xc82075e210, 0xc820222060)
	/ext-go/1/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
	/ext-go/1/src/github.com/cockroachdb/cockroach/util/net.go:127 +0x31f
```

As a note, the alter table statement is accepted and reports as succeeded.  The panic gets triggered by the select statement.

--- a/sql/backfill.go
+++ b/sql/backfill.go
@@ -71,6 +71,7 @@ func (ids indexesByID) Swap(i, j int) {
 func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *roachpb.Error {
 	var droppedColumnDescs []ColumnDescriptor
 	var droppedIndexDescs []IndexDescriptor
+	var newColumnDescs []ColumnDescriptor
 	var newIndexDescs []IndexDescriptor
 	// Mutations are applied in a FIFO order. Only apply the first set
 	// of mutations.
@@ -84,8 +85,7 @@ func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *ro
 		case DescriptorMutation_ADD:
 			switch t := m.Descriptor_.(type) {
 			case *DescriptorMutation_Column:
-				// TODO(vivek): Add column to new columns and use it
-				// to fill in default values.
+				newColumnDescs = append(newColumnDescs, *t.Column)
 
 			case *DescriptorMutation_Index:
 				newIndexDescs = append(newIndexDescs, *t.Index)
@@ -105,12 +105,21 @@ func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *ro
 	// TODO(vivek): Break these backfill operations into chunks. All of them
 	// will fail on big tables (see #3274).
 
-	// Delete the entire dropped columns.
-	// This used to use SQL UPDATE in the past to update the dropped
-	// column to NULL; but a column in the process of being
-	// dropped is placed in the table descriptor mutations, and
-	// a SQL UPDATE of a column in mutations will fail.
-	if len(droppedColumnDescs) > 0 {
+	defaultExprs, err := p.makeDefaultExprs(newColumnDescs)
+	if err != nil {
+		return roachpb.NewError(err)
+	}
+
+	nonNullableColumn := ""
+	// Add the column to the table descriptor.
+	for _, columnDesc := range newColumnDescs {
+		if columnDesc.DefaultExpr == nil && !columnDesc.Nullable {
+			nonNullableColumn = columnDesc.Name
+		}
+		tableDesc.AddColumn(columnDesc)
+	}
+
+	if len(droppedColumnDescs) > 0 || nonNullableColumn != "" || len(defaultExprs) > 0 {
 		// Run a scan across the table using the primary key.
 		start := roachpb.Key(MakeIndexKeyPrefix(tableDesc.ID, tableDesc.PrimaryIndex.ID))
 		// Use a different batch to perform the scan.
@@ -119,6 +128,15 @@ func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *ro
 		if pErr := p.txn.Run(batch); pErr != nil {
 			return pErr
 		}
+
+		if nonNullableColumn != "" {
+			for _, result := range batch.Results {
+				if len(result.Rows) > 0 {
+					return roachpb.NewErrorf("column %s contains null values", nonNullableColumn)
+				}
+			}
+		}
+
 		for _, result := range batch.Results {
 			var sentinelKey roachpb.Key
 			for _, kv := range result.Rows {
@@ -127,6 +145,12 @@ func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *ro
 					// ID. Strip off that suffix to determine the prefix shared with the
 					// other keys for the row.
 					sentinelKey = stripColumnIDLength(kv.Key)
+
+					// Delete the entire dropped columns.
+					// This used to use SQL UPDATE in the past to update the dropped
+					// column to NULL; but a column in the process of being
+					// dropped is placed in the table descriptor mutations, and
+					// a SQL UPDATE of a column in mutations will fail.
 					for _, columnDesc := range droppedColumnDescs {
 						// Delete the dropped column.
 						colKey := keys.MakeColumnKey(sentinelKey, uint32(columnDesc.ID))
@@ -134,6 +158,26 @@ func (p *planner) backfillBatch(b *client.Batch, tableDesc *TableDescriptor) *ro
 							log.Infof("Del %s", colKey)
 						}
 						b.Del(colKey)
+					}
+					// Add the new columns and backfill the values.
+					for i, expr := range defaultExprs {
+						if expr == nil {
+							continue
+						}
+						col := newColumnDescs[i]
+						colKey := keys.MakeColumnKey(sentinelKey, uint32(col.ID))
+						d, err := expr.Eval(p.evalCtx)
+						if err != nil {
+							return roachpb.NewError(err)
+						}
+						val, err := marshalColumnValue(col, d, p.evalCtx.Args)
+						if err != nil {
+							return roachpb.NewError(err)
+						}
+						if log.V(2) {
+							log.Infof("CPut %s -> %v", colKey, val)
+						}
+						b.CPut(colKey, val, nil)
 					}
 				}
 			}

--- a/sql/testdata/alter_table
+++ b/sql/testdata/alter_table
@@ -128,3 +128,69 @@ INSERT INTO t VALUES (4, 1)
 
 statement error duplicate key value \(d\)=\(1\) violates unique constraint \"t_d_key\"
 INSERT INTO t VALUES (5, 1)
+
+# Add a column with no default value
+statement ok
+ALTER TABLE t ADD COLUMN x DECIMAL
+
+# Add a non NULL column with a default value
+statement ok
+ALTER TABLE t ADD COLUMN y DECIMAL NOT NULL DEFAULT (DECIMAL '1.3')
+
+# Add a non NULL column with no default value
+statement error column q contains null values
+ALTER TABLE t ADD COLUMN q DECIMAL NOT NULL
+
+statement ok
+ALTER TABLE t ADD COLUMN z DECIMAL DEFAULT (DECIMAL '1.4')
+
+statement ok
+INSERT INTO t VALUES (11, 12, DECIMAL '1.0')
+
+statement ok
+INSERT INTO t (a, d) VALUES (13, 14)
+
+statement ok
+INSERT INTO t (a, d, y) VALUES (21, 22, DECIMAL '1.0')
+
+statement ok
+INSERT INTO t (a, d) VALUES (23, 24)
+
+statement ok
+INSERT INTO t VALUES (31, 32)
+
+statement ok
+INSERT INTO t (a, d, x, y, z) VALUES (33, 34, DECIMAL '2.0', DECIMAL '2.1', DECIMAL '2.2')
+
+query IITTT colnames
+SELECT * FROM t
+----
+a   d     x     y     z
+1   NULL  NULL  1.3   1.4
+2   NULL  NULL  1.3   1.4
+3   NULL  NULL  1.3   1.4
+4   1     NULL  1.3   1.4
+11  12    1     1.3   1.4
+13  14    NULL  1.3   1.4
+21  22    NULL  1     1.4
+23  24    NULL  1.3   1.4
+31  32    NULL  1.3   1.4
+33  34    2     2.1   2.2
+
+# Subsequent operations succeed because the table is empty
+statement ok
+CREATE TABLE tt (a INT PRIMARY KEY)
+
+statement ok
+ALTER TABLE tt ADD COLUMN q DECIMAL NOT NULL
+
+statement ok
+ALTER table tt ADD COLUMN r DECIMAL
+
+query TTBT colnames
+SHOW COLUMNS FROM tt
+----
+Field  Type     Null   Default
+a      INT      false  NULL
+q      DECIMAL  false  NULL
+r      DECIMAL  true   NULL

